dtls_debug.c

```diff
@@ -345,46 +382,51 @@ void dtls_dsrv_log_addr(log_t level, const char *name, const session_t *addr)
 #ifndef WITH_CONTIKI
 void
 dtls_dsrv_hexdump_log(log_t level, const char *name, const unsigned char *buf, size_t length, int extend) {
-  static char timebuf[32];
-  FILE *log_fd;
   int n = 0;
+#ifndef DTLS_CONSTRAINED_STACK
+  char message[DTLS_DEBUG_BUF_SIZE];
+#endif /* ! DTLS_CONSTRAINED_STACK */
+  int len;
 
   if (maxlog < (int) level)
     return;
 
-  log_fd = level <= DTLS_LOG_CRIT ? stderr : stdout;
-
-  if (print_timestamp(timebuf, sizeof(timebuf), time(NULL)))
-    fprintf(log_fd, "%s ", timebuf);
-
-  if (level <= DTLS_LOG_DEBUG)
-    fprintf(log_fd, "%s ", loglevels[level]);
-
+#ifdef DTLS_CONSTRAINED_STACK
+  dtls_mutex_lock(&static_mutex);
+#endif /* DTLS_CONSTRAINED_STACK */
   if (extend) {
-    fprintf(log_fd, "%s: (%zu bytes):\n", name, length);
-
+    snprintf(message, sizeof(message), "%s: (%zu bytes):\n", name, length);
     while (length--) {
-      if (n % 16 == 0)
-        fprintf(log_fd, "%08X ", n);
-
-      fprintf(log_fd, "%02X ", *buf++);
+      len = strlen(message);
```

Comment:
If the hex dump message, including newlines, is assembled  ahead before calling the handler, it my frequently truncate the output.
Should that be indicated by "..." at the end of the output?
Or should the handler be called per line (maybe adding an additional parameter for follow up lines)?

Reply:
The external logging handler may also be truncating length - which may even affect a line by line mode.  My thoughts here that in the general case it does not really matter if information is lost, bu if space is needed to give the detail in the longer lines, then DTLS_DEBUG_BUF_SIZE can be redefined.  Furthermore, if `extend `is set as a parameter, then the information is not output over multiple lines.

I do like the concept though of "..." indicating something is missing. Code updated and pushed.

Reply:
Instead of calling `strlen()` multiple times, I would prefer to calculate the available space from the return value of `snprintf()`. (Constant instead of linear complexity).